### PR TITLE
sql: allow datestyle and intervalstyle to be set by default

### LINKED ISSUE #69352: sql: remove need for DateStyle/IntervalStyle_enabled
Implementation of this RFC: https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20211220_DateStyle_IntervalStyle_Default.md

- [x] detect interval/date::string conversions in computed columns and partial indexes, and convert them to immutable functions (see https://github.com/cockroachdb/cockroach/pull/68074 as well as https://github.com/cockroachdb/cockroach/issues/68202)
- [ ] remove the session variables

Epic CRDB-10462

Jira issue: CRDB-9555

### PR DESCRIPTION
**Pull request by rafiss**
fixes https://github.com/cockroachdb/cockroach/issues/69352

Release justification: high benefit and low risk change.

Release note (sql change): The session variables datestyle_enabled and
intervalstyle_enabled, and the cluster settings
sql.defaults.datestyle.enabled and sql.defaults.intervalstyle.enabled no
longer have any effect. When the upgrade to v22.1 is finalized, all of
these settings are effectively always interpreted as "true."

Release note (backward-incompatible change): The volatilities of cast
operations between strings and interval/timestamp types has changed from
immutable to stable. This means that these cast operations can no longer
be used in computed columns or partial index definitions. Instead, the
builtin functions parse_interval, parse_date, parse_time, parse_timetz,
parse_timestamp, or to_char can be used. When finalizing the upgrade to
v22.1, CockroachDB will automatically rewrite any computed columns or
partial indexes that use the affected casts to use the new builtin
functions.

### CODE REVIEW

**Review by mgartner — APPROVED** [review]
Very nice! :lgtm: 

This should have the `backport-22.1` label correct?

> Reviewed 18 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78357#-:-MywDdPFBDw3nARNJSIw:b-d1eymm)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/datestyle_intervalstyle_mixed_version, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78357#-MywD20o193PpAuaNiD_:-MywD20o193PpAuaNiDa:bex0icn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda2a38bf035e8d377c82330e5a531aadee94c97/pkg/sql/logictest/testdata/logic_test/datestyle_intervalstyle_mixed_version#L23)):*
> ```
> 
> statement error setting DateStyle is not enabled
> set datestyle = 'dmy'
> ```

nit: uppercase `SET` here are below

Also, should there be a test that this succeeds after `SET datestyle_enabled = true`? Same for interval style.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78357)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/logictest/testdata/logic_test/datestyle_intervalstyle_mixed_version, line 23 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78357#-MywD20o193PpAuaNiD_:-Mywc3SM15KjS5QsHnMt:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cda2a38bf035e8d377c82330e5a531aadee94c97/pkg/sql/logictest/testdata/logic_test/datestyle_intervalstyle_mixed_version#L23)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: uppercase `SET` here are below

Also, should there be a test that this succeeds after `SET datestyle_enabled = true`? Same for interval style.
</blockquote></details>

done

File: pkg/sql/vars.go [hunk 1]
```diff
@@ -2008,6 +2046,26 @@ func init() {
 				return p.setRole(ctx, local, u)
 			},
 		},
+		{
+			name: `role`,
+			fn: func(ctx context.Context, p *planner, local bool, s string) error {
```
Comment (otan) [inline pkg/sql/vars.go:2051]: nit: delete this?

**Review by otan — APPROVED** [review]

File: pkg/sql/vars.go (hunk 1 above)
Comment (rafiss) [reply pkg/sql/vars.go:2051]: whoops, yes

**Comment by rafiss** [comment]
flake was from https://cockroachlabs.slack.com/archives/CJ0H8Q97C/p1648137907442409

tftrs!

bors r=otan,mgartner

**Pull request merged**